### PR TITLE
feat: add delete dialog command

### PR DESCRIPTION
**Pull request by sebastianbuechler**
Tried to add for https://github.com/Stacked-Org/stacked/issues/878

TODO:

- [x] Delete folder in `lib/ui/dialogs/` for respective dialog
- [x] Unregister dialog in `lib/app/app.dart` from dialogs parameter
- [x] Unregister dialog in `app/app.dialogs.dart`

Tested it locally with a new project, added a dialog, then removed it again. Any other tasks here to do?

Includes minor typo fix for `delete_view_command.dart`

### CODE REVIEW

File: lib/src/commands/delete/delete_dialog_command.dart [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  /// Args:
+  ///   outputPath (String): The path to the output folder.
+  Future<void> deleteDialog({String? outputPath}) async {
+    /// Deleting the dialog folder.
+    String directoryPath = _templateService.getTemplateOutputPath(
+      inputTemplatePath: 'lib/ui/dialogs/generic',
+      name: argResults!.rest.first,
+      outputFolder: outputPath,
+    );
+    await _fileService.deleteFolder(directoryPath: directoryPath);
```
Comment (ferrarafer) [inline lib/src/commands/delete/delete_dialog_command.dart:93]: When dialog does not exist, is going to throw a `PathNotFoundException`. Please, catch that exception on `run` method and show a descriptive message, or validate dialog exists before proceed.

Comment (ferrarafer) [inline lib/src/commands/delete/delete_dialog_command.dart:93]: Should also delete Dialog model.

File: lib/src/commands/delete/delete_dialog_command.dart [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        stackTrace: s.toString(),
+      ));
+    }
+  }
+
+  /// It deletes the dialog files
+  ///
+  /// Args:
+  ///   outputPath (String): The path to the output folder.
+  Future<void> deleteDialog({String? outputPath}) async {
```
Comment (ferrarafer) [inline lib/src/commands/delete/delete_dialog_command.dart:86]: - I prefer this function to be private
- Avoid using `argResults` directly on this function, add `dialogName` to the parameters, same we are doing with `outputPath` to be consistence.

File: lib/src/commands/delete/delete_dialog_command.dart [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+      outputFolder: outputPath,
+    );
+    await _fileService.deleteFolder(directoryPath: directoryPath);
+  }
+
+  /// It removes the dialog from [app.dart]
+  ///
+  /// Args:
+  ///   outputPath (String): The path to the output folder.
+  Future<void> removeDialogFromDependency({String? outputPath}) async {
```
Comment (ferrarafer) [inline lib/src/commands/delete/delete_dialog_command.dart:100]: - I prefer this function to be private
- Avoid using `argResults` directly on this function, add `dialogName` to the parameters, same we are doing with `outputPath` to be consistence.

**Review by ferrarafer — CHANGES_REQUESTED** [review]
@sebastianbuechler great PR, thanks for your collaboration. I just made a few requests for changes but overall was very good. Please, try to make the suggested changes and we will merge this PR.

File: lib/src/commands/delete/delete_dialog_command.dart (hunk 1 above)
Comment (sebastianbuechler) [reply lib/src/commands/delete/delete_dialog_command.dart:93]: Did not see that the test files for the dialogs were placed in the viewmodel test folder and forgot to delete them.

Comment (sebastianbuechler) [reply lib/src/commands/delete/delete_dialog_command.dart:93]: Did you mean something like this? Should I add it to the rest of the delete commands as well?

File: lib/src/commands/delete/delete_dialog_command.dart (hunk 2 above)
Comment (sebastianbuechler) [reply lib/src/commands/delete/delete_dialog_command.dart:86]: Done

File: lib/src/commands/delete/delete_dialog_command.dart (hunk 3 above)
Comment (sebastianbuechler) [reply lib/src/commands/delete/delete_dialog_command.dart:100]: Done

**Comment by sebastianbuechler** [comment]
@ferrarafer I also did the adjustment for private functions and parameter addition for the delete commands for service and view so it's consistent.

File: lib/src/commands/delete/delete_dialog_command.dart (hunk 1 above)
Comment (sebastianbuechler) [reply lib/src/commands/delete/delete_dialog_command.dart:102]: I just saw the failing `dart analyze` in the pipeline. Seems that the `PathNotFoundException` is not compatible with the current project constraints of '>=2.17.0 <3.0.0' as it's introduced only in 2.19.0:
https://github.com/dart-lang/sdk/blob/dc30f815f18a4ca641c9206c8895365569ce3644/sdk/lib/io/file.dart#L1005-L1014

File: lib/src/services/pub_service.dart [hunk 4]
```diff
@@ -36,6 +37,8 @@ class PubService {
   /// Checks whether or not has the latest version for `stacked_cli` package
   /// installed on the system.
   Future<bool> hasLatestVersion() async {
+    locator<ColorizedLogService>()
+        .stackedOutput(message: 'Before get current version', isBold: true);
```
Comment (ferrarafer) [inline lib/src/services/pub_service.dart:41]: Why did you add this? 🤔  The function should only check if has or not the latest version. No output needed.

File: lib/src/commands/delete/delete_dialog_command.dart [hunk 5]
```diff
@@ -82,30 +87,46 @@ class DeleteDialogCommand extends Command with ProjectStructureValidator {
...
     );
     await _fileService.deleteFolder(directoryPath: directoryPath);
+
+    //Delete test file for dialog
+    final filePath = _templateService.getTemplateOutputPath(
+      inputTemplatePath: kDialogEmptyTemplateGenericDialogModelTestPath,
+      name: dialogName,
+      outputFolder: outputPath,
+    );
+    await _fileService.deleteFile(filePath: filePath);
```
Comment (ferrarafer) [inline lib/src/commands/delete/delete_dialog_command.dart:110]: You need to verify if the model exists before deletion because we can create dialogs without model, in that case is going to throw an error and halt the command execution.

`PathNotFoundException: Cannot delete file, path = 'my_app/test/viewmodels/warning_dialog_model_test.dart' (OS Error: No such file or directory, errno = 2)`

File: lib/src/commands/delete/delete_dialog_command.dart (hunk 1 above)
Comment (ferrarafer) [reply lib/src/commands/delete/delete_dialog_command.dart:102]: Please, update to 2.19.0.

```
environment:
  sdk: ">=2.19.0 <3.0.0"
```

Comment (ferrarafer) [reply lib/src/commands/delete/delete_dialog_command.dart:102]: > Did you mean something like this? Should I add it to the rest of the delete commands as well?
> 
> https://github.com/Stacked-Org/cli/blob/92bf4617363c663bb023e72f6aa6dd9b33e026cf/lib/src/commands/delete/delete_dialog_command.dart#L75-L83

Yes, you can improve it like below also logging the exception to the analytics.

```
_log.error(message: e.toString());
unawaited(_analyticsService.logExceptionEvent(
      runtimeType: e.runtimeType.toString(),
      message: message,
));
```

File: lib/src/services/pub_service.dart [hunk 6]
```diff
@@ -21,7 +22,7 @@ class PubService {
     for (var package in packages) {
       if (!package.contains(ksStackedCli)) continue;
 
-      version = package.split(' ').last;
+      version = package.split(' ').elementAt(1);
```
Comment (ferrarafer) [inline lib/src/services/pub_service.dart:25]: Why do we need to change this? 🤔

**Review by ferrarafer — CHANGES_REQUESTED** [review]
Sorry for the delay, more changes requested. Thanks for your collaboration, really appreciated.

File: lib/src/services/pub_service.dart (hunk 4 above)
Comment (sebastianbuechler) [reply lib/src/services/pub_service.dart:41]: Forgot to remove it, thanks.

File: lib/src/services/pub_service.dart (hunk 6 above)
Comment (sebastianbuechler) [reply lib/src/services/pub_service.dart:25]: It's fixing this issue: https://github.com/Stacked-Org/stacked/issues/979 But If you want I can do a separate PR. I just needed it to actually test my PR, because I was not able to register the local package in pub otherwise.

File: lib/src/commands/delete/delete_dialog_command.dart (hunk 5 above)
Comment (sebastianbuechler) [reply lib/src/commands/delete/delete_dialog_command.dart]: The exception seems to be only about the test file of the model. The model gets deleted indirectly by deleting the whole dialog folder. I adjusted the code by checking first, if the test file exists:
 

```dart
final fileExists = await _fileService.fileExists(filePath: filePath);
if (fileExists) {
  await _fileService.deleteFile(filePath: filePath);
}
```

File: lib/src/services/pub_service.dart (hunk 6 above)
Comment (sebastianbuechler) [reply lib/src/services/pub_service.dart]: I removed it so we can finish this PR.

**Review by ferrarafer — APPROVED** [review]
Looks good to me, thank you! There is an issue but not because of your code, is an issue that was already on the code, I'm going to fix it on another branch.
